src/coap_gnutls.c

```diff
@@ -770,6 +774,98 @@ setup_pki_credentials(gnutls_certificate_credentials_t *pki_credentials,
     }
     break;
 
+  case COAP_PKI_KEY_PEM_BUF:
+    if (setup_data->pki_key.key.pem_buf.public_cert &&
+        setup_data->pki_key.key.pem_buf.public_cert_len &&
+        setup_data->pki_key.key.pem_buf.private_key &&
+        setup_data->pki_key.key.pem_buf.private_key_len) {
+      gnutls_datum_t cert;
+      gnutls_datum_t key;
+      int alloced_cert_memory = 0;
+      int alloced_key_memory = 0;
+
+      cert.size = setup_data->pki_key.key.pem_buf.public_cert_len;
+      if (setup_data->pki_key.key.pem_buf.public_cert[cert.size-1] != '\000') {
+        /* Need to allocate memory, rather than just copying pointers across */
+        alloced_cert_memory = 1;
+        cert.data = gnutls_malloc(cert.size + 1);
+        if (!cert.data) {
+          coap_log(LOG_ERR, "gnutls_malloc failure\n");
+          return GNUTLS_E_MEMORY_ERROR;
+        }
+        memcpy(cert.data, setup_data->pki_key.key.pem_buf.public_cert,
+               cert.size);
+        cert.data[cert.size] = '\000';
+        cert.size++;
+      }
+      else {
+        /* To get around const issue */
+        memcpy(&cert.data,
+             &setup_data->pki_key.key.pem_buf.public_cert, sizeof(cert.data));
+      }
+      key.size = setup_data->pki_key.key.pem_buf.private_key_len;
+      if (setup_data->pki_key.key.pem_buf.private_key[key.size-1] != '\000') {
```

Comment:
Here, `key.size > 0` is not guaranteed as well.

Reply:
Same logic comment as above.

Reply:
Correct.